planner/core/find_best_task.go

```diff
@@ -846,7 +845,105 @@ func (ds *DataSource) convertToTableScan(prop *property.PhysicalProperty, candid
 	return task, nil
 }
 
-func (ts *PhysicalTableScan) addPushedDownSelection(copTask *copTask, stats *property.StatsInfo) {
+// pushDownSelAndResolveVirtualCols push some filters down to coprocessor and resolve virtual columns.
+//	1. push down filters and check if there is virtual column.
+//	2. substitute virtual columns in TableScan's Schema to corresponding physical columns.
+//	3. substitute virtual columns in TableScan's filters and push them down.
+//	4. add a Projection upon this DataSource.
+func (ds *DataSource) pushDownSelAndResolveVirtualCols(copTask *copTask, path *accessPath, stats *property.StatsInfo) (t task) {
+	// step 1
+	t = copTask
+	if copTask.indexPlan != nil {
+		ds.addPushedDownSelection(copTask, path)
+	}
+	if copTask.tablePlan == nil { // don't need to handle virtual columns in IndexScan
+		return
+	}
+	ts := copTask.tablePlan.(*PhysicalTableScan)
+	if !ds.hasVirtualCol {
+		ds.addPushedDownTableScan(copTask, ts, stats)
+		return
+	}
+
+	// step 2
+	ds.substituteVirtualColumns(ts)
+
+	// step 3
+	for i, expr := range ts.filterCondition {
+		ts.filterCondition[i] = expression.ColumnSubstitute(expr, ds.virtualColSchema, ds.virtualColExprs)
+	}
+	var cantBePushed []expression.Expression
+	_, ts.filterCondition, cantBePushed = expression.ExpressionsToPB(ds.ctx.GetSessionVars().StmtCtx, ts.filterCondition, ds.ctx.GetClient())
+	ds.addPushedDownTableScan(copTask, ts, stats)
+	if len(cantBePushed) > 0 { // for filters cannot be pushed down, we add a root Selection to handle them
+		sel := PhysicalSelection{Conditions: cantBePushed}.Init(ds.ctx, stats)
+		t = sel.attach2Task(copTask)
+	}
+
+	// step 4
+	projExprs := make([]expression.Expression, 0, len(ds.Schema().Columns))
+	for _, c := range ds.Schema().Columns {
+		projExprs = append(projExprs, expression.ColumnSubstitute(c, ds.virtualColSchema, ds.virtualColExprs))
+	}
+	proj := PhysicalProjection{Exprs: projExprs}.Init(ds.ctx, stats)
```

Comment:
```sql
mysql>create table t (a int,b int as (a+1), c int, index idx(b));
Query OK, 0 rows affected
Time: 0.009s
mysql>desc select b,c from t ignore index (idx) where b>1 and c>1;
+---------------------+----------+------+------------------------------------------------------------+
| id                  | count    | task | operator info                                              |
+---------------------+----------+------+------------------------------------------------------------+
| Projection_9        | 1111.11  | root | cast(plus(test.t.a, 1)), test.t.c                          |
| └─Selection_7       | 1111.11  | root | gt(cast(plus(test.t.a, 1)), 1)                             |
|   └─TableReader_8   | 1111.11  | root | data:Selection_6                                           |
|     └─Selection_6   | 1111.11  | cop  | gt(test.t.c, 1)                                            |
|       └─TableScan_5 | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
+---------------------+----------+------+------------------------------------------------------------+
```

As you can see, `cast(plus(test.t.a, 1)` will be calculated twice, both in `Selection_7` and `Projection_9`. Can we do projection first?

Reply:
Good idea, let me have a try.

Reply:
ok